`torch.nn.modules.LazyModuleMixin` and `torch.nn.LazyLinear` (Shape Inference II)

emcastillo opened the pull request:
Retake on #40493 after all the feedback from @albanD 

This PR implements the generic Lazy mechanism and a sample `LazyLinear` layer with the `UninitializedParameter`.

The main differences with the previous PR are two; 
Now `torch.nn.Module` remains untouched.
We don't require an explicit initialization or a dummy forward pass before starting the training or inference of the actual module. Making this much simpler to use from the user side.

As we discussed offline, there was the suggestion of not using a mixin, but changing the `__class__` attribute of `LazyLinear` to become `Linear` once it's completely initialized. While this can be useful, by the time being we need `LazyLinear` to be a `torch.nn.Module` subclass since there are many checks that rely on the modules being instances of `torch.nn.Module`.
This can cause problems when we create complex modules such as
```
class MyNetwork(torch.nn.Module):
    def __init__(self):
        super(MyNetwork, self).__init__()
        self.conv = torch.nn.Conv2d(20, 4, 2)
        self.linear = torch.nn.LazyLinear(10)
    def forward(self, x):
        y = self.conv(x).clamp(min=0)
        return self.linear(y)
```
Here, when the __setattr__ function is called at the time LazyLinear is registered, it won't be added to the child modules of `MyNetwork`, so we have to manually do it later, but currently there is no way to do such thing as we can't access the parent module from LazyLinear once it becomes the Linear module. (We can add a workaround to this if needed).

TODO:

Add convolutions once the design is OK
Fix docstrings

Commit:
Add `LazyModuleMixin` and `LazyLinear`

emcastillo commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,43 @@ def __reduce_ex__(self, proto):
+        if dtype is None:
+            dtype = self.data.dtype
+        self.data = torch.empty(shape, device=device, dtype=dtype)
+        self.__class__ = Parameter
```
This allows us to cleanly avoid dummy forward passes, as for example the optimizers hold a reference to a parameter that will be materialized in the forward pass of the training

emcastillo replied:
That's a good point, maybe we can just set the tensor as property and any access to it will throw an error :)

albanD replied:
@dzhulgakov I don't think we should expect an optimizer that takes `.data_ptr` to work.
Even a `model.to("cuda")` would break such optimizer.

dzhulgakov replied:
Fair point about data_ptr.

There's actually a bigger problem that optimizers with additional data would initialize in the constructor and try to access the shape of the parameter: https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py#L44 . It might not be a huge deal, because if I read the code correctly - it doesn't work with `.to("cuda")` even today (i.e. one has to move the model before creating the optimizer). But it'd be nice to make `shape` on the parameter to throw before it's initialized.

ngimel replied:
But when can optimizer be safely initialized in this case? After moving model to a target device and making a first iteration, because before that parameters wouldn't have shapes?

emcastillo replied:
Actually, in this case, adagrad is not accessing `shape` on the Uninitialized parameter, `full_like` is implemented directly in ATen and is returning an empty tensor :D.

albanD replied:
But that's not correct right? because when we try to use that state with the the initialized version afterwards it will fail no?

emcastillo replied:
Yes, that's actually what I meant but my comment was poorly redacted. It won't hard error in the constructor, but it will error in the step or produce inconsistent results. I added a explicit check to avoid this.

emcastillo commented on torch/jit/_recursive.py:
```diff
@@ -518,6 +518,10 @@ def check_module_initialized(mod):
                            .format(torch.typename(type(mod))))
+    for param in mod.parameters():
+        if isinstance(param, torch.nn.parameter.UninitializedParameter):
+            raise RuntimeError("'{}' has uninitialized parameters or buffers, did you forget to run a forward pass"
```
This needs more debugging ad it might not be needed

albanD replied:
What's the plan to check this? Shall we ask someone from the jit side to take a look?

dzhulgakov replied:
cc @suo

emcastillo replied:
would be nice 😅

suo replied:
This seems fine—from what I understand, after the first forward we'll have fully initialized the parameters and the `LazyModule` will be indistinguishable from a regular `nn.Module`, right? If so, then it's fine to just error and ask the user to run forward.

albanD commented on test/test_nn.py:
```diff
@@ -12342,6 +12343,114 @@ def local_backward_hook(m, input, output):
+        module = LazyModule()
+        module.register_parameter('test_param', UninitializedParameter())
+        module.test_param.materialize(10)
+        module.cuda()
```
Shouldn't the cuda be before `materialize` here?

albanD commented on test/test_nn.py:
```diff
@@ -12342,6 +12343,114 @@ def local_backward_hook(m, input, output):
+
+class TestLazyModules(TestCase):
+
+    def test_lazy_module_parameter(self):
```
It would be interesting to have something similar to this for Linear/LazyLinear.
In particular you can add a test to ensure that you can load a Linear into a LazyLinear.

albanD commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,116 @@
+
+
+class LazyModuleMixin:
+    """A mixin for modules that lazily initialize buffers and parameters.
```
The location for this docstring might not be great as users won't check this doc in general but only the `LazyLinear` one (or corresponding Module they use).

If we only do `LazyLinear` for now, we should just move this to that docstring.
When we add convolution later, maybe this shared part can be moved to a note in the doc that each lazy module we implement refers to.


Also an important part of this doc should be when we need to do a dummy forward pass or not.
In particular it is not required before creating optimizers or serializing but it is unclear what should be done when the user wants to do custom initialization of their weights, or use DataParallel, or DistributedDataParallel, or reparametrization like `nn.utils.weight_norm()`.

yf225 replied:
Agreed, having the doc on when we need to do a dummy forward pass is super important for setting expectations :)

albanD reviewed (commented):
Thanks for updating the PR!

We had a chance to sit down with @mruberry and @ngimel to discuss the PR.
We do think it is much "safer" than the previous one and is looking quite good. See the inline comments below.

One question I would have is if it would be possible to remove the  LazyModuleMixin from the `__class__` after we ran the hook and everything is initialized? After this point, we don't need the custom loading logic at all right?

emcastillo commented on test/test_nn.py:
```diff
@@ -12342,6 +12343,114 @@ def local_backward_hook(m, input, output):
+        self.assertEqual(module.test_param, torch.ones((5, 5))) 
+
+        # Uninitialized parameters are left unchanged
+        new_module = LazyModule()
```
This was a typo, I copypasted the old tests and there still some work left to be done, thanks a lot for pointing it out

emcastillo commented:
Thanks for the awesome review as always, will work on it and send changes soon 😇

yf225 commented:
Thanks a lot for working on this @emcastillo! This would be really useful as we can take advantage of this lazy-init concept for one of our internal PyTorch library as well.

One question I have is: do you know how it would interact with the .to() API? would it error out if the parameters are not initialized yet?

emcastillo commented:
We allow moving the parameters to a different device before materializing them, so it is safe to call .to before the initialization :)
They will be instantiated in the correct device

emcastillo commented on test/test_nn.py:
```diff
@@ -12342,6 +12343,114 @@ def local_backward_hook(m, input, output):
+        module.register_parameter('test_param', UninitializedParameter())
+        self.assertTrue(module.has_uninitialized_params())
+
+        new_module = LazyModule()
```
Yes, the `new_module` above was a typo, it should be `module`

emcastillo commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,116 @@
+            if key in state_dict:
+                input_param = state_dict[key]
+                # Detect Unitialized parameters or buffers
+                if isinstance(input_param, UninitializedParameter):
```
The current error that happens is as you said, shape mismatch.
```
 size mismatch for weight: copying a param with shape torch.Size([0]) from checkpoint, the shape in current model is torch.Size([10, 5]).```

However, if we override `shape` in `LazyParameter` to throw an error, this message can be more descriptive.
What do you think?

emcastillo commented on torch/nn/modules/linear.py:
```diff
@@ -176,4 +177,20 @@ def extra_repr(self) -> str:
         )
 
+
+class LazyLinear(LazyModuleMixin, Linear):
```
I don't see a way to enforce this other than the tests ...

albanD replied:
Sounds good.

emcastillo commented:
I think I addressed all the comments. 
Now we forbid to call `param.shape`, fixed the tests, docs and also make the module to change its class to `Linear` once it's initialized.

yf225 commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,50 @@ def __reduce_ex__(self, proto):
+
+    Uninitialized Parameters holds empty tensors that can be moved
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration. Moreover, Unitialized parameters
```
minor nit:
```suggestion
    they do it directly with the right configuration. Moreover, uninitialized parameters
```

emcastillo replied:
Thanks!

yf225 commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,116 @@
+            raise RuntimeError('module {} has not been fully initialized'.format(self._get_name()))
+        self._initialize_hook.remove()
+        if self.cls_to_become is not None:
+            self.__class__ = self.cls_to_become
```
Curious: would it work in general if the lazy module class is actually the class I want to use after the module has been initialized? (e.g. for a library where all modules are lazy and there is no "eager" module to replace with)

I might be overlooking some test cases, but it would be really awesome to highlight that this is possible :)

emcastillo replied:
if you dont override this parameter, it will remain as a lazy module😃

emcastillo replied:
Great idea! I just added it :}

albanD commented on torch/nn/modules/module.py:
```diff
@@ -967,10 +967,12 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
-                if len(param.shape) == 0 and len(input_param.shape) == 1:
+                # We access data.shape because for uninitialized parameters `param.shape`
+                # is forbidden
+                if len(param.data.shape) == 0 and len(input_param.data.shape) == 1:
```
I know this is "ok" because these are `Parameter` but I really don't like to see `.data` added here.
Given the current code, you could use `.size()` or `.ndim()` that you don't block.

dzhulgakov replied:
I think it's actually fine to just do `isinstance(param, UninitializedParam)` here, it justifies intent much better and I don't think adds too much cyclical dependency

albanD commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,50 @@ def __reduce_ex__(self, proto):
+    @property
+    def shape(self):
+        raise RuntimeError(
+            'Can\'t access the shape of an uninitialized parameter. '
```
If this is to make sure that loading the state dict of an LazyModule into a regular one is failing, maybe add a note in the error message that if this happens during load_state_dict, this is because you are trying to load an uninitialized parameter into an initialized one.

dzhulgakov replied:
I think it's a generic error message to catch the problems in user code too.

Actually, for completeness - shall we override `size` too? `ndim` I guess is fine as the rank usually doesn't change

albanD replied:
It is a generic message but also to catch bad loading (from the test, this is the error that is looked for).
I still think it's worth adding a hint in the error if we expect it to be a major reason for this to be raised. Otherwise, people loading models and getting this error will be very confused.

albanD commented on test/test_nn.py:
```diff
@@ -13019,6 +13020,130 @@ def local_backward_hook(m, input, output):
+    def test_optimizer_pass(self):
+        module = LazyModule()
+        module.register_parameter('test_param', UninitializedParameter())
+        optim = torch.optim.SGD(module.parameters(), lr=0.0)
```
I think this is a very good test and we should do it for all the optimizers (with basic settings).
We want to make sure that PRs that update the optimizers don't break this kind of logic.

albanD reviewed (commented):
This looks quite good for me.
One question I still have beyond the inline comments is how does it interact with DataParallel and DistributedDataParallel? If they don't we should raise an error or at least document it very clearly if raising an error is tricky.

dzhulgakov reviewed (commented):
This looks really good! Some really minor comments, but I think it's close to landing

ngimel commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,116 @@
+    one; especially, ``module.cuda()`` may cause the initialization to run on
+    a GPU.
+    The default value of lazy buffers and parameters are ``torch.Tensor([])``
+    and ``UninitializedParameter()``, respectively.
```
I feel like this is missing examples of how one would use LazyModule in common scenarios, e.g. what if one wants to do custom initialization? DataParallel? Reparameterization such as weight_norm or spectral_norm? Common case where none of the above is needed?

emcastillo replied:
`DataParallel` and `weight_norm` or `spectral_norm` do not work without a dummy pass first, or unless we change their code to deal with uninitialized parameters.  I think we can detect these situations easily and just raise an error.
asking the user to do a dummy forward on everything that tries to access the uninitialized parameter data.

ngimel replied:
It's ok that they don't work, the ask here is for documentation and examples (as @albanD and @yf225 mentioned)

emcastillo commented:
Hi, I've been debugging the use of DataParallel with the module and I saw that we require a explicit dummy forward pass.
This is because when the module is replicated, all the parameters are copied as tensors so we lose the ability to magically convert an uninitialized parameter into a parameter and it remains as an empty tensor which can't be materialized.

https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/replicate.py#L144-L146

Solutions to this are:
1. Error when a Lazy module is passed to DataParallel and ask the user to do a dummy forward pass (this requires the user to know how the input tensor will be split across the modules).

2. Do the dummy forward pass in the DataParallel itself before replicating the module, if the module is lazy and has uninitialized parameters we do this in the forward call. (Cons, tricky and too magical)

albanD commented:
3. sounds tricky because you will need to keep the autograd link between the parameter and the buffer which can be tricky.
I think that 1. is a safe alternative for now.

emcastillo commented:
Pushed some more changes with the reviews!
Thanks a lot

emcastillo commented:
exactly, I added a explicit check to this optimizer.
This was a (poor) explanation of why raising an error in a parameter doesnt work here 😞

albanD commented on test/test_nn.py:
```diff
@@ -13019,6 +13020,160 @@ def local_backward_hook(m, input, output):
+                      torch.optim.RMSprop, torch.optim.LBFGS]
+
+        def run_step(module, optim):
+            optim = torch.optim.SGD(module.parameters(), lr=0.0)
```
nit you always recreate a SGD here instead of using the optim that was given.

emcastillo replied:
lol, I am stupid

albanD commented on test/test_nn.py:
```diff
@@ -13019,6 +13020,160 @@ def local_backward_hook(m, input, output):
+            self.assertIsInstance(optim.param_groups[0]['params'][0], UninitializedParameter)
+            module.test_param.materialize(10)
+            self.assertIsInstance(optim.param_groups[0]['params'][0], Parameter)
+            self.assertNotIsInstance(optim.param_groups[0]['params'][0], UninitializedParameter)
```
It would be nice to add a random grad and do a step here just to make sure that the buffers get properly created:

```python
for p in module.parameters():
  p.grad = torch.rand_like(p)

optim.step()
```

albanD commented on torch/nn/modules/module.py:
```diff
@@ -986,6 +988,7 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                                       'whose dimensions in the checkpoint are {}, '
                                       'an exception occurred : {}.'
                                       .format(key, param.size(), input_param.size(), ex.args))
+                    raise ex
```
This should never raise!

emcastillo replied:
some testing code that I forgot to delete ...

albanD commented on torch/nn/modules/module.py:
```diff
@@ -965,18 +965,20 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                # This is used to avoid copying uninitialized parameters into
+                # non-lazy modules, since they dont have the hook to do the checks
+                # in such case, it will error when accessing the .shape attribute.
+                is_param_lazy = isinstance(param, torch.nn.parameter.UninitializedParameter)
```
So you rely on the `.copy_()` below failing and providing the error message here?

emcastillo replied:
When accessing the lazy attribute .shape it will hard-error, the rest of errors are already covered by the hook.
If I can't access a parameter data, it becomes quite tricky to cover all the cases with simple code.

mruberry replied:
Like @albanD I'm a little confused by this code, too. Is there a simpler or more direct way to write this check, possibly by testing if the param is lazy vs. testing if it's not lazy twice?

mruberry replied:
Alternatively, if lazy modules don't support state dicts maybe this code can just be removed.

mruberry replied:
Following our other discussions, I think this is OK for now, but we should revisit this approach in the near future.

albanD commented on torch/optim/adagrad.py:
```diff
@@ -40,6 +40,8 @@ def __init__(self, params, lr=1e-2, lr_decay=0, weight_decay=0, initial_accumula
         for group in self.param_groups:
             for p in group['params']:
+                if isinstance(p, torch.nn.parameter.UninitializedParameter):
+                    raise ValueError('torch.optim.AdaGrad does not support torch.nn.parameter.UninitializedParameter')
```
nit: You can add: "Make sure to run the dummy forward before creating the optimizer."

emcastillo replied:
yes, good point!

albanD reviewed (commented):
Looks good.
What are the current error that we get for weight_norm/spectral_norm? If they are not nice, we can add a check there to raise a proper error [here](https://github.com/pytorch/pytorch/blob/1c6ace87d127f45502e491b6a15886ab66975a92/torch/nn/utils/weight_norm.py#L28) for example.

Similarly for DDP? We can add a check there to make sure it fails with a useful error message.

ngimel commented:
> Similarly for DDP? We can add a check there to make sure it fails with a useful error message.
DP should probably error out when attempting to replicate, and that's fine. For DDP there can be a nasty error mode where initial broadcast with uninitialized parameters succeeds, but then each process initializes parameter in its own way. We should have something preventing that (may be it is prevented already, and initial broadcast fails? That would be good, but needs tests).

emcastillo commented:
Added the missing checks, tests for DDP, norms and corrected the majority of the comments

Thanks a lot!

albanD reviewed (approved):
Thanks for taking the time to add all the tests.
LGTM

yf225 commented on torch/optim/adagrad.py:
```diff
@@ -40,6 +40,10 @@ def __init__(self, params, lr=1e-2, lr_decay=0, weight_decay=0, initial_accumula
+                if isinstance(p, torch.nn.parameter.UninitializedParameter):
+                    raise ValueError(
+                        '`torch.optim.AdaGrad` does not support torch.nn.parameter.UninitializedParameter'
+                        'Make sure to run the dummy forward before creating the optimizer.')
```
Is it correct that we only need to check this in `Adagrad` and not other optimizers? For user-created optimizers in the wild, currently is there any mechanism to hard-fail and mention the need of dummy forward pass?

In general I am thinking whether we should put more of these checks into the `UninitializedParameter` class methods rather than in the use sites (probably except for a few cases where we have to catch these errors early)

albanD replied:
> In general I am thinking whether we should put more of these checks into the UninitializedParameter class methods rather than in the use sites

I think that from the user experience side, having a clear error telling them what to do when they try to use it is quite important and is a good reason to put these errors at the use site.

> Is it correct that we only need to check this in Adagrad and not other optimizers?

Yes. There are tests for the other optimizers to make sure they work fine.

> For user-created optimizers in the wild, currently is there any mechanism to hard-fail and mention the need of dummy forward pass?

Unfortunately no. And because they could be creating states in any way they want, it is hard to say how to catch these. Hopefully, people following the example from our basic optimizers should be lazily initializing buffers and it will work fine.

dzhulgakov replied:
I wonder whether we should override `__torch_function__` for UninitializedParameter class. That should give good error messages in attempting to use this op in any of the torch operations. E.g. in `torch.full_like` below. Have we considered that?

emcastillo replied:
I jus tried this but it was pretty tricky to filter accessing to `param.grad_fn` or `param.data.device` without giving errors...

emcastillo replied:
pushed :)

yf225 replied:
@emcastillo curious with the `__torch_function__ ` change, do we still need to keep the `UninitializedParameter` check in other places in this PR? (happy with either decision, just wanted to make sure it's an intentional one :)

emcastillo replied:
I think that a hard-error in dataparallel, or the weight norms, is more user-friendly than just failing in a random torch function that the user is not aware of, so I think both checks can co-exist.

ngimel commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,148 @@
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    buffers outside of the constructor (``__init__``).
```
buffers -> parameters

ngimel commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,148 @@
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" forward once to initialize all parameters in case
+    that you need to access the weights before the actual real ``forward`` pass.
+
```
putting dummy initialization example here instead of below would be logical

ngimel commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,148 @@
+    dummy forward call in the following scenarios.
+
+
+    >>> lazy_module = LazyLinear(10)
```
this code sample seems out of place here, it's not a scenario requiring explicit forward

ngimel commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,148 @@
+    >>> module_dp = torch.nn.DataParallel(lazy_module)
+
+    * Module reparametrization using :func:`torch.nn.utils.weight_norm` or :func:`torch.nn.utils.spectral_norm`
+    >>> torch.nn.utils.weight_norm(module)
```
module -> lazy_module?

ngimel replied:
Also, where's an example of non-default weight initialization? It can be said to be implicit in "access data or properties", but probably makes sense to also mention it explicitly.

ngimel commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,148 @@
+    * Access an :class:`UnitializedParameter` data or properties
+    >>> lazy_module.weight.shape
+
+    * Use an Optimizer that access Parameters data in their constructor 
```
access -> accesses

emcastillo commented:
Thanks for the feedback, added the corrections to the docstrings!

emcastillo commented:
I just added the `__torch_function__` check, it makes hard-erroring automatic so it should be able to detect many overlooked situations and provide meaningful error messages. I am not very confident about the implementation so feedback is welcomed.

I think that if this change is OK, we should be able to merge this soon.

yf225 commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,83 @@ def __reduce_ex__(self, proto):
+            torch.Tensor.double,
+            torch.Tensor.cuda,
+            torch.Tensor.cpu,
+            torch._has_compatible_shallow_copy_type]
```
curious: does `Tensor.to` work on `UninitializedParameter`?

emcastillo replied:
yes, when we `materialize` the parameter will be created in the device where the uninitialized parameter was moved to 😊

yf225 replied:
thanks @emcastillo ! curious should we allow other dtype conversion APIs such as `long()` as well? (I might also be missing other tensor APIs that we could enable on UninitializedParameter :)

albanD commented:
Note that the CI failures are real.

dzhulgakov commented on torch/jit/_recursive.py:
```diff
@@ -540,6 +540,10 @@ def check_module_initialized(mod):
     if not hasattr(mod, '_parameters'):
         raise RuntimeError("'{}' has not been initialized, did you forget to call 'super()'?"
                            .format(torch.typename(type(mod))))
+    for param in mod.parameters():
```
so the test failures can be seen in the full error message: https://circleci.com/api/v1.1/project/github/pytorch/pytorch/8056474/output/105/0?file=true&allocation-id=5f7d8d4f7099d65d3111d1e5-0-build%2F48B29293

I think it comes from this line:
```
Oct 07 11:29:49 Traceback (most recent call last):
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 259, in wrapper
Oct 07 11:29:49     fn()
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 90, in new_test_method
Oct 07 11:29:49     return_value = old_test_method(self, *arg, **kwargs)
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 132, in test_forward_async
Oct 07 11:29:49     for remote_module in self._create_remote_module_iter(dst_worker_name):
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 103, in _create_remote_module_iter
Oct 07 11:29:49     scripted_remote_module = torch.jit.script(remote_module)
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/jit/_script.py", line 898, in script
Oct 07 11:29:49     obj, torch.jit._recursive.infer_methods_to_compile
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/jit/_recursive.py", line 350, in create_script_module
Oct 07 11:29:49     check_module_initialized(nn_module)
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/jit/_recursive.py", line 543, in check_module_initialized
Oct 07 11:29:49     for param in mod.parameters():
Oct 07 11:29:49   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/distributed/nn/api/remote_module.py", line 266, in parameters
Oct 07 11:29:49     "Method ``parameters`` not supported for RemoteModule. Please use ``remote_parameters`` instead."
Oct 07 11:29:49 ValueError: Method ``parameters`` not supported for RemoteModule. Please use ``remote_parameters`` instead.
```

I'm not sure what's the right condition to put here - maybe check that it's not RemoteModule? @suo or @mrshenli  probably know

albanD replied:
Note that this is failing internally as well because we call this module with things that are "almost" nn.Module but don't have a `.paramters()` method.
Maybe you should check if there is a `._parameters` dict and if so traverse it?

emcastillo replied:
Thanks, I will fix this.

dzhulgakov commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,88 @@ def __reduce_ex__(self, proto):
+            torch.Tensor.to,
+            torch._has_compatible_shallow_copy_type]
+        # method-wrapper is to detect access to Tensor properties that are 
+        # wrapped in descriptors
```
what exactly does it mean? Is it stuff like `torch.shape`?

emcastillo replied:
when calling an attribute that is accessed through a [descriptor](https://docs.python.org/3/howto/descriptor.html) the func object that we get is a bit weird

dzhulgakov commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,88 @@ def __reduce_ex__(self, proto):
+        )
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
```
thanks, I think the coverage with torchfunction will be much better!

this looks good to me, but @ezyang or @hameerabbasi are experts here and probably should take a look

hameerabbasi commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,88 @@ def __reduce_ex__(self, proto):
+        # method-wrapper is to detect access to Tensor properties that are 
+        # wrapped in descriptors
+        if func in allowed_methods or func.__class__.__name__ == 'method-wrapper':
+            return super().__torch_function__(func, types, args)
```
```suggestion
            return super().__torch_function__(func, types, args, kwargs)
```

emcastillo replied:
This causes the following error on some methods
```
takes no keyword arguments
```

emcastillo replied:
Added some workaround as I couldn't find a cleaner way ...

hameerabbasi commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,88 @@ def __reduce_ex__(self, proto):
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        allowed_methods = [
```
Nit: Can we make this a class attribute and a set, so it isn't initialized each time this method is called?

hameerabbasi reviewed (approved):
I've reviewed the `__torch_function__` specific changes. Please feel free to tag me if there's something else you think I should review.

gchanan commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,78 @@ def extra_repr(self) -> str:
+    become a :class:`torch.nn.Linear` dropping all the lazy initialization
+    logic.
+
+    Be sure to run "dummy" forward once to initialize all parameters in case
```
I don't understand what the punctuation is trying to get across.  Why is dummy quoted but real isn't?  Why is the first forward not double `` but the second one is?

emcastillo replied:
I think it should be consistent now, thanks!

hameerabbasi commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,95 @@ def __reduce_ex__(self, proto):
+        # method-wrapper is to detect access to Tensor properties that are 
+        # wrapped in descriptors
+        if func in cls._allowed_methods or func.__class__.__name__ == 'method-wrapper':
+            if kwargs is None:
+                # in some calls such as is_floating_point or attribute access
+                # even kwargs=None will result in a `takes no keyword arguments` error
+                return super().__torch_function__(func, types, args)
+            else:
+                return super().__torch_function__(func, types, args, kwargs)
```
```suggestion
            if kwargs is None:
                kwargs = {}
            return super().__torch_function__(func, types, args, kwargs)
```

hameerabbasi replied:
This is the standard way to do it.

emcastillo commented:
I think i've covered all the requests and CI issues. Hope we can land it soon!

emcastillo commented:
Wonder what is the internal error 🤔

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,78 @@ def extra_repr(self) -> str:
+        bias: If set to ``False``, the layer will not learn an additive bias.
+            Default: ``True``
+
+    Attributes:
```
I worry about the docs duplication between this and the original Linear module. Is there a way to avoid the redundant documentation so that updates to the linear module don't have to be reflected here, too?

emcastillo replied:
I added the `nn.lazy.LazyModuleMixin` to the documentation and put a pointer to it in the linear docs

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,78 @@ def extra_repr(self) -> str:
+    become a :class:`torch.nn.Linear` dropping all the lazy initialization
+    logic.
+
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
```
I think this should reference the lazy documentation to avoid redundancy and stay up to date.

mruberry replied:
Per above comment, this sentence can be removed.

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,78 @@ def extra_repr(self) -> str:
+    r"""A :class:`torch.nn.Linear` module with lazy initialization.
+
+    This module will initialize the `weight` and `bias` parameters after
+    the first call to ``forward`` is done. After that the module will
```
It'd be interesting to see how the weight and bias are constructed from the input here.

emcastillo replied:
I extended the docstring to reflect this

mruberry commented on torch/jit/_recursive.py:
```diff
@@ -559,6 +559,13 @@ def check_module_initialized(mod):
+    if not hasattr(mod, 'remote_parameters'):
+        for param in mod.parameters():
+            if isinstance(param, torch.nn.parameter.UninitializedParameter):
+                raise RuntimeError("'{}' has uninitialized parameters or buffers, did you forget to run a forward pass"
```
"pass" -> "pass?"

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
```
This is misleading, since regular modules can also create and initialize or modify parameters outside of their constructors.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+
+    Unlike regular modules, subclasses of this module can initialize
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
```
Again, nothing would prevent a regular module from doing this, either. The value of this class is not in the lazy construction of the parameters.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+        self._load_hook = self._register_load_state_dict_pre_hook(self._lazy_load_hook)
+        self._initialize_hook = self.register_forward_pre_hook(self._infer_parameters)
+
+    def _save_to_state_dict(self: _LazyProtocol, destination, prefix, keep_vars):
```
Things have changed a little since this feature was first proposed. What's the value of these modules having a state dict?

emcastillo replied:
As @yf225 suggested, there are modules that will remain as lazy ones because there are no actual non-lazy modes that they can be converted to after the initialization. These kinds of modules should support serialization as well.

mruberry replied:
That's extremely interesting. Sorry @yf225 I missed your question about this. Could you elaborate on your thinking? I'm worried about the complexities of serializing lazy modules.

emcastillo replied:
I completely understand the issue with saving the state of a lazy module, a lazy module that has not been initialized has no actual state, but the state of a non-lazy module can be loaded turning it into a non-lazy module.

I think that forbidding serializing a LazyModule (by throwing an exception when accessining an `UninitializedParameter` is fine. In this case a module that inherits from the lazy mixin but has no UninitializedParameters could still work. 
However, ideally, when the module its materialized, it should remove the LazyMixin and hooks from its `__class__.__bases__` and become a regular nn.Module instance. @yf225 what are your thoughts on this?

@albanD do you agree with raising an error when the state dict is requested?

mruberry replied:
Per our other discussion, let's not change this code now, but we may want to revisit serialization in the future.

mruberry commented on torch/nn/utils/weight_norm.py:
```diff
@@ -36,7 +36,10 @@ def apply(module, name: str, dim: int) -> 'WeightNorm':
 
         weight = getattr(module, name)
-
+        if isinstance(weight, UninitializedParameter):
```
I'm not in love with these norms being aware of uninitialized parameters. Would you it be possible to have the lazy linear modules throw this error on their own? 

For example, the lazy linear module could define two empty tensors, _bias and _weight, and then when bias or weight is requested the module itself could throw an error like this, I think?

emcastillo replied:
Sorry, right now I don't understand how adding two extra parameters can help to detect these kinds of errors.

mruberry replied:
I'm probably not being very clear about this. Maybe we should leave it for future work.

emcastillo replied:
I understood it from your example in another comment. However, this affects the current design so we should discuss this better. I will reply in that comment.

mruberry replied:
Sounds good. Let's leave this as-is for now.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
```
We should be clear in the documentation that this is a new feature whose design and implementation are likely to change. Before the 1.8 release (~3 months from now) we should have a design review to determine if we think the feature is stable. 

Following that logic, I think that when a lazy module is constructed it should throw the same warning.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
```
I'd like to motivate this initial forward pass better. Effectively, the module's parameters don't exist before the first forward pass is run, and some PyTorch functionality assumes those parameters already exist (and why wouldn't they?). To initialize the lazy modules, then, a "dry run" with inputs of a "representative size", dtype and device should be performed. 

Detailing the dry run is a little tricky. In PyTorch networks can be very dynamic and have multiple branches. A single initializing dry run may not call foward() on each module in a network. If lazy modules appear in two of a network's branches, for example, then two initializing dry runs would be necessary, and the inputs for each dry run must be constructed to trigger the different paths through the network.

You can try to address this complexity or suggest, for now at least, that users not try to use lazy modules with networks with multiple branches.

emcastillo replied:
I tried to add some comments from your reply to this docstring.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    >>> lazy_module = torch.nn.LazyLinear(10)
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
```
Let's use language like, "dry run initializes the lazy modules".

It would be helpful to show the before/after structure of a simple network here. For example, a MLP with two linear layers and a ReLU in-between them. Also, this should make clear that the input for the dry run must be of an appropriate size. It would also help if a transform, like putting the modules on CUDA, was applied to the network so users understand when to perform transforms like that, and then to show an optimizer being attached to the network, too.

emcastillo replied:
Added the example.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
+    >>> assert type(lazy_module) is torch.nn.Linear
+
+    Note that lazy modules cannot validate if the shape is correct during
```
I appreciate what this paragraph is trying to say but this concept is complicated enough that you'd need more time and an example to explain it.

This is really a note on initialization. An example like the one I recommend above could be helpful. If both layers are lazy then the network will be initialized just like the non-lazy version. If only the first layer is lazy, however, then the initialization will be different. I think you'd need to show that here to make this point clearly.

emcastillo replied:
Added the example as you suggested :)

mruberry replied:
Thank you!

At this point in the documentation the purpose of lazy modules as well as their basic usage pattern has been explained and demonstrated. The remainder of the doc needs to mention serialization and initialization. Serialization needs to be introduced. You could do something like this:


"Lazy modules can be serialized with a state dict like other modules. For example:

```
<brief code snippet show a lazy linear's state dict and loading from that state dict>
```

Lazy modules can also load regular :class:\`torch.nn.Parameter\`s, which replace their :class:\`torch.nn.UninitializedParameter\`s:

```
<brief code snippet showing loading initialized parameters from a lazy module after a dry run or a regular linear layer>
```

Note, however, that lazy modules cannot validate that the shape of parameters they load is correct.
"

emcastillo replied:
Ah, I need to update this, I missed this comment

emcastillo replied:
Added examples :)

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+    deserialization.  Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one; especially, ``module.cuda()`` may cause the initialization to run on
```
This note about CUDA is not lazy module specific and can be removed.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,195 @@
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
```
Maybe we should just make the initial dry run part of the lazy pattern? This will be even more important if lazy modules aren't serializable. Also, if we think about PyTorch networks being modular then it's hard to know how someone else might use them, making an initial dry run an important part of the program.

emcastillo replied:
lazy modules are currently serializable. 
What do you mean by making the initial dry run part of the lazy pattern? 
adding some kind of explicit `initialize` method to a lazy module?

mruberry replied:
> lazy modules are currently serializable.

I understand. But should they be? I'm not sure I understand the benefit.

> What do you mean by making the initial dry run part of the lazy pattern?
> adding some kind of explicit `initialize` method to a lazy module?

Not exactly. I mean that we tell people the way to use modules is to put the network on the appropriate device and dtype, perform a dry run, then continue modifying the network or accessing its parameters (by, for example, attaching an optimizer to it). Anything else could throw an error message.

emcastillo replied:
Yes, more or less this is how it currently works, but the dry_run is only mandatory in certain cases. If I understood this correctly, we should make the dry run mandatory regardless of the use case. It is a change that I agree with as it makes everything explicit.

mruberry replied:
You're completely correct. However, following our other discusisons, for now let's remove this note. In the future we can be more explicit about when a dry run isn't required.

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,78 @@ def extra_repr(self) -> str:
+        super().__init__(0, out_features, bias)
+        self.weight = UninitializedParameter()    
+
+    def reset_parameters(self) -> None:
```
I'm curious if we can make it so that:

- attempts to access the parameters of a lazy module throws an error
- the forward pass has the logic to "materialize" AND register the params AND convert the module to not be lazy

emcastillo replied:
> attempts to access the parameters of a lazy module throws an error
When initializing the parameters we need to explicitly access them, I think it's not easy at all to discriminate whether do we have or not legit accesses to a parameter.

> the forward pass has the logic to "materialize" AND register the params AND convert the module to not be lazy

Sorry, can you provide some kind of example? I don't currently see how this could be done.

mruberry replied:
I understand. I'm asking what if we broke the property that references to an UninitializedParameter are supported.

For example, in linear we have weight and bias. Could we do something like the following:

- Delete the weight and bias parameters and make attempts to access them throw an error
- Create dummy _weight and _bias UninitializedParameters that are just empty tensors that can have their dtype and device change, but trying to edit them otherwise is an error
- Have the LazyModule mixin cause attempts to access the parameters to throw an error
- Then, on forward (or pre forward), the weight and bias parameters are created as usual with shapes derived from the input data and dtype and device derived from _weight and _bias
- The _weight and _bias UninitializedParameters are removed
- The LazyModule class is removed

I think that would give us a regular linear layer? Put another way, I'm wonder if we can create a module that will let the network it's in work with .to(), but it requires a forward pass before anything else can be done to it (like registering an optimizer).

emcastillo replied:
This is quite a design change with several implications,
Right now an `UninitializedParameter` changes it's data and `__class__` when it is materialized.
https://github.com/pytorch/pytorch/blob/6de9ebd676bed3ff3ed269767a6a8a2ab2a503a2/torch/nn/parameter.py#L96-L97

But for the materialization to happen, we need to access the parameter itself, so we can't forbid this access in the current pattern.
https://github.com/pytorch/pytorch/blob/6de9ebd676bed3ff3ed269767a6a8a2ab2a503a2/torch/nn/modules/linear.py#L254

With this trick, if some code such as optimizers were holding a reference to an `UninitializedParameter`, that reference will still be valid after the materialization. However, this was a trick to avoid the explicit dry run in simple cases, so your proposal actually makes the design more robust if we forbid accessing a state dict and make the dry run mandatory.
I would like to hear @albanD and @yf225 opinions on this. I am fine with either of the approaches as they have both their cons/pros.

mruberry replied:
Great points. See my comment below, too. Maybe we wait on this decision. But let me check with the team.

mruberry replied:
Per discussion elsewhere, let's not make this change now. We'll revisit it later.

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,93 @@ def __reduce_ex__(self, proto):
+
+
+class UninitializedParameter(Parameter):
+    r"""A parameter that can be lazily initialized.
```
It's not that the parameter can be lazy initialized so much as the uninitialized parameter represents a parameter that is not yet initialized.

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,93 @@ def __reduce_ex__(self, proto):
+class UninitializedParameter(Parameter):
+    r"""A parameter that can be lazily initialized.
+
+    Uninitialized Parameters holds empty tensors that can be moved
```
Start by briefly describing how UninitializedParameters are like Parameters, then describe how they are different: basically that many operations supported on normal parameters are not supported on UninitializedParameters.

emcastillo replied:
Updated ccomment

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,93 @@ def __reduce_ex__(self, proto):
+    they do it directly with the right configuration. Moreover, Uninitialized parameters
+    become full fledged :class:`torch.nn.Parameter` after they are initialized.
+    """
+    _allowed_methods = [
```
This is a great idea. Should dunder getattr throw an error if the requested attribute isn't supported, too?

emcastillo replied:
Yes, we should definitely do that. However, I can't think of attributes that we need to forbid the access to right now 😅

mruberry replied:
We can leave that for future work. Maybe a comment like "TODO: have dunder getattr throw an error when an unsupported attribute is requested" is all we need for now?

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,93 @@ def __reduce_ex__(self, proto):
+    become full fledged :class:`torch.nn.Parameter` after they are initialized.
+    """
+    _allowed_methods = [
+        torch.Tensor.__hash__,
```
What if dunder hash, size, and copy_ weren't supported?

emcastillo replied:
These are needed to support manipulating the `UninitializedParameters` for the materialization and loading a state_dict. I think this needs to be documented.

mruberry replied:
Are they still needed if state dicts are unsupported? Also, if "materialization" constructs new parameters then I don't think we need these, either?

emcastillo replied:
materialize currently changes the parameter data in place. 
Also loading a state dict should be supported and this triggers a `copy_` call,'
`__hash__` is used when iterating through the module parameters, so it's actually being called in a bunch of different places.

mruberry replied:
OK, sounds good for now.

mruberry reviewed (changes requested):
Hey @emcastillo!

I'm really impressed with the thoughtfulness that has gone into this PR, and that thoughtfulness has encouraged me to take the time to think carefully about this design and what improvements may be available. I've made some comments/suggestions reflecting that thinking.

Most importantly, I'd like to be clear that this feature is new and likely to change before the 1.8 release as we continue to develop it. Beyond that, I think there's an opportunity to simplify the design of lazy modules and simultaneously clarify their presentation. I think these possible simplifications are a natural next step that follows on the good work you've done.

I'm really curious to hear your thoughts on the proposed changes and which direction you'd like to go in. I'd also like to see this PR landed sooner or later, so I'll be sure to respond to your updates quickly.

albanD commented on torch/jit/_recursive.py:
```diff
@@ -559,6 +559,13 @@ def check_module_initialized(mod):
 
+    # This is to avoid importing torch.distributed.nn
+    if not hasattr(mod, 'remote_parameters'):
+        for param in mod.parameters():
```
This does not help for the internal tests as you're still accessing `.parameters()` on a Module-like object that don't have such a function.
As mentioned before, I think you need to iterate over the `._parameters` directly here.

ezyang commented:
Just a gentle reminder to all of the reviewers here that emcastillo doesn't have push permissions, so someone (@albanD, judging from the import to Phabricator) should shepherd this to completion. This includes agreeing that not all comments have to be resolved on *this* PR and that we can evolve the feature further in master.

emcastillo commented:
Thanks for the incredible review @mruberry!!
There are some points I don't fully understand yet so some reply to my comments would be really appreciated.
Thanks a lot again

mruberry commented:
> Thanks for the incredible review @mruberry!!
> There are some points I don't fully understand yet so some reply to my comments would be really appreciated.
> Thanks a lot again

Happy to respond. If some of these issues are too complicated to deal with now then we can also consider merging this PR and continuing to develop it together in future PRs.

emcastillo commented:
Thanks for the follow-up!!!

From the discussion above there are mainly two points that need to be agreed upon asap since they change the design of the feature itself.

1) Forbid serialization
I am ok with not allowing a lazy module state dict to be retrieved when there are uninitialized parameters, however I think we should allow loading a non-lazy state dict into a lazy module fully initializing it.

2) Explicit dry-run
Right now, a dry-run is required only in certain cases, there are error checks placed in several parts of the code to make the user aware of this. However as a design philosophy and to make the pattern consistent I think it is not bad to request the lazy forward always. With this, some ideas such as https://github.com/pytorch/pytorch/pull/44538#discussion_r503719650
could be applied, reducing the need for explicit error checks in several places such as the weight/spectral normalizations.

@mruberry @albanD @yf225 @ezyang @ngimel what are your thoughts on these points?

mruberry commented:
> Thanks for the follow-up!!!
> 
> From the discussion above there are mainly two points that need to be agreed upon asap since they change the design of the feature itself.
> 
> 1. Forbid serialization
>    I am ok with not allowing a lazy module state dict to be retrieved when there are uninitialized parameters, however, I think we should allow loading a non-lazy state dict into a lazy module fully initializing it.
> 2. Explicit dry-run
>    Right now, a dry-run is required only in certain cases, there are error checks placed in several parts of the code to make the user aware of this. However as a design philosophy and to make the pattern consistent I think it is not bad to request the lazy forward always. With this, some ideas such as [#44538 (comment)](https://github.com/pytorch/pytorch/pull/44538#discussion_r503719650)
>    could be applied, reducing the need for explicit error checks in several places such as the weight/spectral normalizations.
> 
> @mruberry @albanD @yf225 @ezyang @ngimel what are your thoughts on these points?

I agree with your thinking and appreciate your open-mindedness. I'll coordinate with the team ASAP tomorrow (it's very late here at the moment) and we'll get back to you. If we can't reach a decision immediately then I propose we stick with the current design for this PR and revisit changes like this later.

yf225 commented:
@mruberry @emcastillo For one of the (large) PyTorch domain library we are building internally, we only have lazy-init modules and there is no non-lazy module, because lazy-init modules are awesome and we definitely don't want to have two classes (lazy, non-lazy) per module to avoid confusing users.

For explicit dry-run, we are actually already doing it for the internal library we are migrating to PyTorch, and it's been working fine. We told users that 1. they need to dry run the model before accessing its parameters (e.g. for creating optimizers), 2. only the branch used by the dry run input will be activated, and if they want to use other branches, they need to create a new module instance and dry run it with the right input.

For serialization, my take is that:
1. We shouldn't allow serializing an unmaterialized lazy module (there is probably no need since user can just re-create the lazy module by calling the class constructor with the right arguments).
2. We should allow deserializing a non-lazy state dict into a lazy module. From user POV I think this should be a valid way of materializing the lazy module. If on the other hand we say "lazy module cannot load state dict unless they are materialized", which means we need to dry run the lazy module before calling `load_state_dict()`, it's still okay but just an extra step for users to do.

mruberry commented:
Hey @emcastillo,

@yf225 @albanD @ngimel and I had a chance to sync. Here's what we'd like to propose:

- At this point, code changes only to fix bugs / build issues. @albanD and @yf225 are going to review the internal build issues and make recommendations for how to fix them. 
- I'm going to review the documents again. In particular, we thought we'd simplify them for now by telling users to always do a dry run. In the future we can relax this requirement.
- Once this PR is landed, we can have additional discussions about design changes/improvements. Since this feature will warn users that it's still in development we don't expect this to be a problem.

How does that sound? @yf225 is really eager to start using this work.

emcastillo commented:
Hi, thanks a lot for everything!

I just pushed a commit that should address all the pending stuff, however I fiddled with the sphinx docs so I am not sure if documentation is going to be correctly rendered without broken the references

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,243 @@
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Lazy modules are a special type of modules that can hold
```
These updates are good but this paragraph and the next one now have redundant information. You could do something like this: 

"Modules that lazily initialize parameters, or "lazy modules," derive the shapes of their parameters from the first input(s) to their forward method. Until that first forward they contain  :class:`torch.nn.UninitializedParameter`s that should not be accessed or used, and afterward they contain regular :class:`torch.nn.Parameter`s. Lazy modules are convenient since they don't require computing some module arguments, like the `in_features` argument of a typical :class:`torch.nn.Linear`.

After construction, networks with lazy modules should first be converted to the desired dtype and placed on the desired device. The lazy modules should then be initialized with one or more "dry runs". These "dry runs" send inputs of the correct size, dtype, and device through the network and to each lazy of its lazy modules. After this the network can be used as usual.

For example:

```
class LazyMLP(torch.nn.Module):
    ...    def __init__(self):
    ...        super().__init__()
    ...        self.fc1 = torch.nn.LazyLinear(10)
    ...        self.relu1 = torch.nn.ReLU()
    ...        self.fc2 = torch.nn.LazyLinear(1)
    ...        self.relu2 = torch.nn.ReLU()
    ...
    ...    def forward(self, input):
    ...        x = self.relu1(self.fc1(input))
    ...        y = self.relu2(self.fc2(x))
    ...        return y
>>> # constructs a network with lazy modules
>>> lazy_mlp = LazyMLP()
>>> # transforms the network's device and dtype
>>> # NOTE: these transforms can and should be applied after construction and before any 'dry runs'
>>> lazy_mlp = lazy_mlp.cuda().double()  
>>> lazy_mlp
    LazyMLP(
      (fc1): LazyLinear(in_features=0, out_features=10, bias=True)
      (relu1): ReLU()
      (fc2): LazyLinear(in_features=0, out_features=1, bias=True)
      (relu2): ReLU()
    )
>>> # performs a dry run to initialize the network's lazy modules
>>> lazy_mlp(torch.ones(10, 10, dtype=torch.double, device='cuda'))  
>>> # after initialization, LazyLinear modules become regular Linear modules
>>> lazy_mlp
    MLP(
      (fc1): Linear(in_features=10, out_features=10, bias=True)
      (relu1): ReLU()
      (fc2): Linear(in_features=10, out_features=1, bias=True)
      (relu2): ReLU()
    )
>>> # attaches an optimizer, since parameters can now be used as usual
>>> optim = torch.optim.SGD(lazy_mlp.parameters(), lr=0.01)
```

emcastillo replied:
Thanks!!

Any idea of how to solve this?
```
ct 15 06:33:30 /var/lib/jenkins/workspace/docs/source/nn.rst.rst:383: WARNING: autosummary: stub file not found 'nn.lazy.LazyModuleMixin'. Check your autosummary_generate setting.
Oct 15 06:33:30 /opt/conda/lib/python3.6/site-packages/torch/nn/modules/linear.py:docstring of torch.nn.LazyLinear:36: WARNING: Explicit markup ends without a blank line; unexpected unindent.
Oct 15 06:33:30 looking for now-outdated files... none found```

I am not very good with sphinx 😅

emcastillo replied:
Updated, thanks!!!

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,246 @@
+    :class:torch.nn.UninitializedParameters that should not be accessed
+    or used, and afterward they contain regular :class:torch.nn.Parameters.
+    Lazy modules are convenient since they don't require computing some
+    module arguments, like the in_features argument of a typical :class:torch.nn.Linear.
```
This needs some additional updates and a few formatting corrections because the formatting doesn't copy+paste easily from the Github comments. For example, ":class:torch.nn.Linear" -> ":class:\`torch.nn.Linear\`".

emcastillo replied:
sorry, my mistake :)

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,246 @@
+    configured, as the order of initialization is different from the original
+    one.
+
+    In the example above, if both layers are `LazyLinear` the initialization
```
Now is the bit on initialization. You could do something like this:

"A final caveat when using lazy modules is that the order of initialization of a network's parameters may change, since the lazy modules are always initialized after other modules. This can cause the parameters of a network using lazy modules to be initialized differently than the parameters of a network without lazy modules. For example, if the LazyMLP class defined above had a :class:\`torch.nn.LazyLinear\` module first and then a regular :class:\`torch.nn.Linear\` second, the second module would be initialized on construction and the first module would be initialized during the first dry run."

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,246 @@
+    first layer is initialized in the dry run, resulting in different weights than the non
+    lazy version of the network.
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
```
I think this can now be removed since it's included earlier.

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,246 @@
+    * When dealing with non-default module initialization
+    >>> torch.nn.init.uniform_(lazy_module.weight)
+
+    .. warning:
```
I suggest moving this warning up in the documentation (see my updated comment about the start of these docs).

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,246 @@
+                      'so changes to the API or functionality can happen at any moment.')
+
+    def _save_to_state_dict(self: _LazyProtocol, destination, prefix, keep_vars):
+        r"""Saves module state to `destination` dictionary, containing a state
```
Can we leave this function undocumented for now? I think it's very likely to change in the future.

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,62 @@ def extra_repr(self) -> str:
+class LazyLinear(LazyModuleMixin, Linear):
+    r"""A :class:`torch.nn.Linear` module with lazy initialization.
+
+    In this module, the `weight` and `bias` are of :class:`torch.nn.Uninitialized`
```
The class is UninitializedParameter, not Uninitialized.

mruberry replied:
Instead of re-explaining what a lazy module is I think this can redirect the user to the lazy module documentation itself.

emcastillo replied:
updated

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,62 @@ def extra_repr(self) -> str:
+    class. They will be initialized  after the first call to ``forward`` is done.
+    The parameters are materialized in the `initialize_parameters` method, which
+    will instantiate them with any device or datatype the module is set to.
+    After the dry run, the module will become a :class:`torch.nn.Linear`
```
This last sentence is work keeping, since it's not duplicative with the lazy module documentation, but it needs a small tweak:

"After the first call to forward this module will instantiate its uninitialized parameters and become a regular :class:`torch.nn.Linear` module."

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,62 @@ def extra_repr(self) -> str:
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    Note that lazy modules cannot validate if the shape is correct during
```
This paragraph can be removed because it's covered by the lazy module documentation above.

mruberry commented on torch/nn/modules/linear.py:
```diff
@@ -178,4 +179,62 @@ def extra_repr(self) -> str:
+                :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})` where
+                :math:`k = \frac{1}{\text{in\_features}}`
+
+    .. note:: check the :class:`torch.nn.modules.lazy.LazyModuleMixin` for documentation
```
Remove this note here and add a sentence above pointing to this document.

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,102 @@ def __reduce_ex__(self, proto):
+    Unitialized Parameters are a a special case of :class:`torch.nn.Parameter`
+    where the shape of the data is still unknown.
+
+    Differentely from :class:`torch.nn.Parameter`, uninitialized parameters
```
"Differentely from..." -> "Unlike a ..."

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,102 @@ def __reduce_ex__(self, proto):
+    where the shape of the data is still unknown.
+
+    Differentely from :class:`torch.nn.Parameter`, uninitialized parameters
+    do not hold any data and the access to the data related information such
```
"do not hold any data and the access to the data related information such as shape is forbidden" -> "hold no data and attempting to access some properties, like their shape, will throw a runtime error."

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,102 @@ def __reduce_ex__(self, proto):
+
+    Differentely from :class:`torch.nn.Parameter`, uninitialized parameters
+    do not hold any data and the access to the data related information such
+    as shape is forbidden. The only operations that can be done in a uninitialized
```
"that can be done in a" -> "that can be performed on an"

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,102 @@ def __reduce_ex__(self, proto):
+    Differentely from :class:`torch.nn.Parameter`, uninitialized parameters
+    do not hold any data and the access to the data related information such
+    as shape is forbidden. The only operations that can be done in a uninitialized
+    parameter are, changing the datatype, moving it to a different device and
```
"are, changing the datatype," -> "are changing its dtype,"

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,102 @@ def __reduce_ex__(self, proto):
+    do not hold any data and the access to the data related information such
+    as shape is forbidden. The only operations that can be done in a uninitialized
+    parameter are, changing the datatype, moving it to a different device and
+    `materialize` them.
```
"\`materialize\`" -> "'materializing'" them.

mruberry commented on torch/nn/parameter.py:
```diff
@@ -44,3 +44,102 @@ def __reduce_ex__(self, proto):
+    parameter are, changing the datatype, moving it to a different device and
+    `materialize` them.
+
+    When uninitialized parameters are materialized, they do it directly 
```
I think this paragraph can be removed now.

mruberry commented:
Hey @emcastillo! You're so fast I'm afraid you resolved some of my comments before I could even finish editing them. Please take a close look because I've made some updates and, in a few places, you'll need to be more careful than I was about formatting. 

I believe I've reviewed all the documentation now. I also updated some other conversations with our most recent thinking for clarity. I think after another round of docs revisions we should be in good shape. I'm looking forward to seeing this PR land.

Please let me know if you have additional questions and ping me when you've had an opportunity to review and update the docs.

emcastillo commented:
hahah, I will give another pass :),
Thanks a lot for taking all this time in correcting all these things, I really appreciate all the help.

Commit:
serialization examples

Commit:
fix sphinx

mruberry commented on torch/nn/modules/lazy.py:
```diff
@@ -0,0 +1,276 @@
+    than the parameters of a network without lazy modules.
+    For example, if the LazyMLP class defined above had a :class:`torch.nn.LazyLinear` module
+    first and then a regular :class:`torch.nn.Linear` second, the second module would be
+    initialized on construction and the first module would be initialized during the first dry run."
```
Remove the quotation mark (") at the end of this line.

Commit:
minor fixes

mruberry reviewed (approved):
Looking forward to having the start of lazy modules in PyTorch!

emcastillo commented:
Thanks a lot everyone for all the help and bringing this to completion!
I will start adding convolutions and batch normalization!
